### PR TITLE
FIP-0057: Update gas charging schedule and system limits for FEVM

### PR DESCRIPTION
**Pull request by raulk**
**Discussions:** https://github.com/filecoin-project/FIPs/discussions/588

With the introduction of user programmable smart contracts, we need to ensure that the gas model accurately reflects the cost of on-chain computation. In the current network, users can only trigger course-grained behaviors by submitting messages to built-in actors. However, with the introduction of FEVM, users will be able to deploy new smart contracts with fine-grained control over FVM execution. If we don't carefully tune the gas model to accurately charge for computation, an attacker could carefully construct a contract to take advantage that and slow down or even stop chain validation.

This FIP proposes a few adjustments to the gas charging schedule. Specifically:

- Adjustments to some syscall gas:
    - State read and write costs.
    - Internal send costs.
    - Hashing, signature verification, and randomness.
- Variable charges for memory initialization, copying, and allocation.
- New charges for state-tree reads and writes.
- New charges for actor address resolution.

This FIP also introduces overall memory limits. However, these memory limits currently only include Wasm memories and table elements (not IPLD blocks, wasm _code_, and other metadata).

Additionally, this FIP introduces a maximum block size limit of 1MiB for all newly created IPLD blocks.

Finally, this FIP reduces the maximum recursive call depth limit from 1025 to 1024 to bring it in-line with the initial intentions of this limit and other blockchain VMs.

### CODE REVIEW

**Comment by raulk** [comment]
@Stebalien has been working extensively in this area and he's the best person to take this FIP to completion.

File: FIPS/fip-nnnn-fevm-gas.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+With the introduction of FEVM, users will be able to trigger these operations:
+
+1. Operations that have runtime cost varying in the size of the inputs (bulk memory copies, hashing, etc.).
+2. Syscalls traversing the extern boundary.
+3. Syscalls not traversing the extern boundary.
+4. Sequences of predefined Wasm instructions backing FEVM opcode handlers.
+5. Memory allocation.
+
+We propose gas charges and system limits to Currently, there is a fixed fee but no per-byte charge for hashing, memory copies, fills, etc. Furthermore, all Wasm instructions are priced through execution gas at a flat rate of 4 gas/op with the exception of some control flow instructions. However, recent observations suggest that some incur in no CPU cost, so we propose new gas exemptions for higher accuracy.
```
Comment (maciejwitowski) [inline FIPS/fip-nnnn-fevm-gas.md:43]: Part of the sequence after "system limits to" is missing.

File: FIPS/fip-nnnn-fevm-gas.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+Instructions that operate on variable-sized memory chunks are now charged a per-byte memory copy fee: `{table,memory}.{init,fill,copy,grow}`.
+
+This fee is identical to `p_memcpy_gas_per_byte` as defined in [FIP-0032](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0032.md), currently priced at 500 milligas per byte.
+
+Additionally, these fees are charged for memory and table allocation when the actor is instantiated (started).
+
+#### No-Ops
+
+Costs have been removed from some operations that would have little to no cost on a traditional CPU. These rules are _valid_ under FEVM because the user cannot trigger _specific_ sequences of WebAssembly instructions but must be revisited before introducing full user programmability as some pathologically unoptimized programs may be able to find a way to make these instructions actually take time.
```
Comment (maciejwitowski) [inline FIPS/fip-nnnn-fevm-gas.md:109]: Does "full user programmability" mean "native actors" here? IIUC we should add it as an issue to M2.2 release, right?

**Comment by anorth** [comment]
Can we please have a discussion link? There are technical impact things to discuss here, including impact on built-in actors and core network operations.

**Comment by Stebalien** [comment]
@anorth done.

**Review by anorth — COMMENTED** [review]
Thanks. This looks ok to me as a first draft. However I think the Product Considerations is currently underdone until it includes indicative numbers for the changing costs of core network operations on the built-in actors.

File: FIPS/fip-nnnn-fevm-gas.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
+---
+fip: "<to be assigned>"
```
Comment (kaitlin-beegle) [inline FIPS/fip-nnnn-fevm-gas.md:2]: FIP0057 :)

Comment (raulk) [reply FIPS/fip-nnnn-fevm-gas.md:2]: Updated in the title. @Stebalien can you update in the FIP?

**Review by jennijuju — APPROVED** [review]

**Pull request merged**